Refinery: add interface to kindlyTo-transformation

klees opened the pull request:
Hi everybody,

this proposes a missing piece from our `Refinery` input validation framework, which are the `kindlyTo` transformation to implement [Postels Law of robustness](https://en.wikipedia.org/wiki/Robustness_principle) when dealing with input from other systems.

We had three workshops of round about 4 hours to [discuss the details](https://docu.ilias.de/goto_docu_xpdl_7929.html). Thanks @mjansenDatabay, @chfsx, @alex40724, @swiniker, @Amstutz and @nhaagen. This is what we came up with, and we all thought it would be helpful to see our results in a collected form to see if they are internally coherent.

We also found that we would want to have `optionalInt`, `optionalBool`, ... to match the `null or int`, `null or bool`, ... cases. To keep this digestible I did not include these here.

Since this will be something every ILIAS developer will need to work with at some point I invite everybody to ask questions, express (dis-)agreement or give feedback in general.

Thanks and best regards!

klees committed:
Refinery: add interface to kindlyTo-transformation

klees commented:
Hi JF,

I labeled this for the next Jour Fixe to raise awareness for this proposal. Dear colleagues, please have a look into this and give some feedback. Maybe we can move on here soon.

Thanks!

mjansenDatabay commented:
@klees LGTM! Looks exactly like what we've defined in the workshops.

klees commented:
@mjansenDatabay Phew, seems I managed to copy and paste correctly :sweat_smile:

kergomard commented on src/Refinery/KindlyTo/Group.php:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    /**
+     * Get a kind transformation to a tuple.
```
```suggestion
     * Get a kind transformation to a record.
```
Müsste record sein, oder?

kergomard commented on src/Refinery/KindlyTo/Group.php:
```diff
@@ -0,0 +1,190 @@
+ *
+ * They don't try to mimic PHPs type cast, but instead follow more sophisticated
+ * rules devised in a series of workshops with interested developers from the
+ * community. Thanks Michael Jansen, Fabian Schmid, Alex Killing, Stephan Winicker,
```
```suggestion
 * community. Thanks Michael Jansen, Fabian Schmid, Alex Killing, Stephan Winiker,
```
;-)

klees replied:
sry =(

kergomard commented on src/Refinery/KindlyTo/Group.php:
```diff
@@ -0,0 +1,190 @@
+     * Get a kind transformation to a `DateTimeImmutable`.
+     *
+     * This supports:
+     *   - all formats mentioned in DateTimeInterface, which are proped in a
```
```suggestion
     *   - all formats mentioned in DateTimeInterface, which are probed in a
```
?

kergomard reviewed (approved):
Thank you very much! Just details in the text below.

matthiaskunkel commented:
Jour Fixe, 11 MAY 2020 : Richard asks all developers for feedback to this suggestion and invites to ask him and Michael.

klees commented:
Hi JF,

I just want to announce that I will merge this soon and that my new colleague @lukastocker is working on an implementation that is almost finished. So this could be used soon!

Best regards!

matthiaskunkel commented:
Jour Fixe, 03 AUG 2020 : We highly appreciate this suggestion and accept the PR for trunk.

lukastocker committed:
Refinery KindlyTo - Group.php: Set up references to transformations and test file.

lukastocker committed:
Refinery KindlyTo - Boolean: transform data into boolean and test transformation.

lukastocker committed:
Refinery KindlyTo - Integer: transform data into integer and test transformation.

lukastocker committed:
Refinery KindlyTo - String: transform data into string and test transformation.

lukastocker committed:
Refinery KindlyTo - Float: transform data into float and test transformation.

lukastocker committed:
Refinery KindlyTo - Dictionary: transform data into dictionary and test transformation.

lukastocker committed:
Refinery KindlyTo - List: transform data into a list and test transformation.

lukastocker committed:
Refinery KindlyTo - Tuple: transform data into a tuple and test transformation.

lukastocker committed:
Refinery KindlyTo - Record: transform data into a record and test transformation.

lukastocker committed:
Refinery KindlyTo - DateTimeImmutable: transform specific data into a DateTimeImmutable and test transformation.

klees commented:
Hi @mjansenDatabay,

my colleague @lukastocker implemented this in the meantime. I did some polishing and also reviewed several times, but I would appreciate a third pair of eyes. Would you please review this? Everyone else is welcome as well, of course.

I changed the description in two interfaces slightly. I realized we cannot accept arrays without keys for record and dict, so I removed two lines that said we would wrap non-arrays into arrays for the according transformations. I also understood that we could indeed strip down big arrays to smaller ones when expecting records. Since this interface is not in production currently (so change is not breaking) and the adjustments are small, I would not go to the JF with this again. Feel free to disagree, though.

Thanks a lot and best regards!

mjansenDatabay commented on src/Refinery/KindlyTo/Group.php:
```diff
@@ -0,0 +1,213 @@
+     *   - ints, which will be typecasted to float
+     *   - bools, where true maps to 1.0 and false to 0.0
+     * This doesn't support:
+     *   - "" will be discarded, as well as well as null, because null or empty are not 0
```
`as well as well` should be `as well`

mjansenDatabay commented on src/Refinery/KindlyTo/Transformation/TupleTransformation.php:
```diff
@@ -0,0 +1,87 @@
+<?php
```
Our coding style should be applied to this file.

mjansenDatabay commented on src/Refinery/KindlyTo/Transformation/StringTransformation.php:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * @inheritdoc
+     */
+    public function transform($from)
```
What about a `$from` of type `object`, implementing `__toString`? According to the interface description this should be transformed as well.

I already changed this in my `Review PR`, but maybe there are questions regarding this change?!?

klees replied:
I merged this and added a test for that as well. Thx!

mjansenDatabay commented on src/Refinery/KindlyTo/Transformation/BooleanTransformation.php:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * @inheritdoc
+     */
+    public function transform($from)
```
Can we really expect that the `Multibyte String Functions` are always available in PHP >= 7.3? In class `ilStr` the existence of these functions is always checked by `function_exists`.

klees replied:
Interesting point =)

The [PHP documentation](https://www.php.net/manual/de/function.mb-strtolower.php) says that this should be available starting with PHP >= 4.3 already... So I guess we could just drop the checks in ilStr.

mjansenDatabay commented on src/Refinery/KindlyTo/Transformation/DateTimeTransformation.php:
```diff
@@ -0,0 +1,64 @@
+        }
+
+        if (is_int($from) || is_float($from)) {
+            $UnixTimestamp = strtotime($from);
```
According to https://www.php.net/manual/en/function.strtotime.php `strtotime` accepts only argumens of type `string`. Here, you pass either an `int` or a `float`.

This will result in a `PHP Fatal error:  Uncaught TypeError: strtotime() expects parameter 1 to be string, int given` if strict types are enabled.

The interface description defines: `integers, which will be interpreted as Unix timestamps.`

When reading this, I would expect the following code:

```php
if (is_int($from)) {
    return new \DateTimeImmutable('@' . $from);
}
```

klees replied:
Nice catch. Thanks!

mjansenDatabay replied:
:eyes: FYI: I did not change this, yet. So if you agree, we'll have to change this (you can pass this to me if you like).

klees replied:
Yeah, I know =) I'm currently on it.

mjansenDatabay commented on src/Refinery/KindlyTo/Transformation/TupleTransformation.php:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * @param Transformation[] $transformations;
+     */
+    public function __construct($transformations) {
```
We should add an `array` typehint here.

mjansenDatabay commented on src/Refinery/KindlyTo/Transformation/TupleTransformation.php:
```diff
@@ -0,0 +1,87 @@
+     * @inheritDoc
+     */
+    public function transform($from) {
+        $this->ValueLength($from);
```
We should rearrange these lines end call `ValueLength` after we ensured `$from` is a non empty array.

mjansenDatabay commented on src/Refinery/KindlyTo/Transformation/TupleTransformation.php:
```diff
@@ -0,0 +1,87 @@
+        return $result;
+    }
+
+    private function ValueLength($values) {
```
We should add an `array` typehint here. Furthermore we should use `void` as return type declaration.

I'd also like to rename this method.

mjansenDatabay commented on src/Refinery/KindlyTo/Transformation/RecordTransformation.php:
```diff
@@ -0,0 +1,78 @@
+    /**
+     *@param Transformation[] $transformations
+     */
+    public function __construct($transformations)
```
We should add an `array` typehint here.

mjansenDatabay reviewed (changes requested):
Hello @klees and @lukastocker!

Thanks for implementing the `kindlyTo` group. IMO this is a huge step forward!

I had only minor remarks, maybe we should have a closer look at the `DateTimeTransformation` and `StringTransformation`.
I created a PR for some of my requested changes: https://github.com/klees/ILIAS/pull/50 . 

Best regards,
@mjansenDatabay

Amstutz reviewed (commented):
Hi @lukastocker and @klees

Very neat. Thx very much to you both.

I just reviewed the interface and the tests for correctness and completeness. If both are fine, the middle part (implementation) is not so much of interest to me...

The interface seems fine. Thx a lot for the detailed descriptions and documentation of what is supported and what not!

I would feel somewhat more comfortable about the middle part without looking at it if for the tests you could add:
- [x] Boolean Tests I: "true" and "false" **in all kinds of capitalization** seems missing. 
- [x] Boolean Tests II: "null" or null, since the absence of some data is something else then true or false seems missing
- [x] Boolean Tests III: 1.0 or 0.0, because we don't expect that someone really wants to transmit booleans disguised as floats seems missing
- [x] Boolean Tests IV: Some Failing Transformation Tests for "All other data will be discarded." are missing. For bool, those are imo important, e.g. ensure that 2 or "2" will fail.
- [x] Float Tests I: strings in natural notation matching is partly missing. I would like to see „01“ failing.
- [x] Float Tests II: trimming is supported is missing. I would like to see „ 1 “ passing.
- [x] Float Tests III: Same with Floating Point representation
- [x] Float Tests IV:  delimiters per mill, like 1'000 be discarded is missing
- [x] Integer Tests I: floats, which will be rounded naturally, add one making sure round up is not used.
- [x] Integer Tests II: strings matching \s*(0|(-?[1-9]\d*))\s*, trimming is supported, add one for trimming supported. Add one excluding „01“. Add one for „0“.
- [x] Integer Tests III: add several tests making sure everything else is discarded (see doesn’t support list for int)
- [x] String Tests I:  All other data will be transformed using __toString, just one very basic test.


Not sure about this, take or leave:
- [ ] Maybe change to different named method for the tests, making clear, which support case is tested. E.g. move 'pos_boolean' => ['true', true] to testSupportTrueFalseInclCapitalization.

Thx you both very much!

klees commented:
Hi @Amstutz, hi @mjansenDatabay,

thanks for looking into this and giving such valuable feedback. Special kudos for requesting more tests: I indeed found some problems (and fixed them). Did you know that `NaN !== NaN` is true and NaNness needs to be checked with `is_nan`? Weird stuff, we did good deciding to ban `NaN` =)

Best regards!

mjansenDatabay reviewed (changes requested):
Hello @klees!

Thanks for your changes.

I am still not sure regarding `DateTimeTransformation` and `This supports ... integers, which will be interpreted as Unix timestamps.`
For me, a unix timestamp is a numeric value representing the number of seconds elapsed since midnight proleptic Coordinated Universal Time (UTC) of January 1, 1970. In https://github.com/ILIAS-eLearning/ILIAS/pull/2584/files#diff-5dbc324b6310e86c3ee20fa37956c5cfR49 the code accepts variables of type `int` and `float`, which are then casted to a `string` and finally passed to the constructor of `DateTimeImmutable `.

Let's have a look at the tests:
```php
return [
    'unix_timestamp' => [20200706122305, \DateTimeImmutable::createFromFormat(\DateTimeImmutable::ISO8601, '2020-07-06T12:23:05+0000')]
];
```

**Depending** on the given PHP timezone given in the context of code execution, `(new DateTimeImmutable((string) 20200706122305))->format(\DateTimeImmutable::ISO8601);` will result in something like `2020-07-06T12:23:05+0200`.
If we interprete `20200706122305` **as a unix timestamp** (like described in the interface docs), the result with the same timezone will be `642104-11-23T00:11:45+0000`.

In my opinion `20200706122305` is an integer representation of a datetime string like `2020-07-06 12:23:05`, but **not** a unix timestamp. If we really want to support unix timestamps here, we **must** use:

```php
if (is_int($from)) {
    return new \DateTimeImmutable('@' . (string) $from);
}
```

Clients/Consumers of the `Refinery` will most probably receive an `int` based unix timestamp (when not receiving datetime strings in the [supported formats](https://github.com/ILIAS-eLearning/ILIAS/pull/2584/files#diff-5dbc324b6310e86c3ee20fa37956c5cfR31)). To transform an `int` to a datetime object which is equivalent to `2020-07-06T12:23:05+0200`, the input value (in this case) must be `1594030985` and the call would be like ` $this->transformation->transform(1594030985);`.

Maybe we should have a chat about this.

Best regards,
@mjansenDatabay

klees commented:
@mjansenDatabay No need to chat, you are correct. Thanks a lot!

klees commented:
@mjansenDatabay Do you see any harm in accepting floats as Unix timestamps as well?

mjansenDatabay commented:
@klees No, I don't.

klees commented:
I'm really exited to use this =)
